### PR TITLE
Add schema inference to `clickhouse-obfuscator`

### LINKED ISSUE #39521: `clickhouse-obfuscator`: add schema inference
Don't require the `--structure` argument.

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add schema inference to `clickhouse-obfuscator`, so the `--structure` argument is no longer required. 

Closes #39521

### CODE REVIEW

File: programs/obfuscator/Obfuscator.cpp [hunk 1]
```diff
@@ -1287,32 +1291,63 @@ try
...
+
+        ReadBufferIterator read_buffer_iterator = [&]()
+        {
+            read_buffer_from_fd = std::make_unique<ReadBufferFromFileDescriptor>(STDIN_FILENO);
+            auto read_buf = std::make_unique<PeekableReadBuffer>(*read_buffer_from_fd);
+            read_buf->setCheckpoint();
+            return read_buf;
+        };
+
+        auto context_const = WithContext(context).getContext();
```
Comment (evillique) [inline programs/obfuscator/Obfuscator.cpp:1313]: Not sure whether it is the best way to get ContextPtr out of ContextMutablePtr

File: programs/obfuscator/Obfuscator.cpp [hunk 2]
```diff
@@ -1287,32 +1291,57 @@ try
...
+
+        for (size_t i = 0, size = structure_vals.size(); i < size; i += 2)
+        {
+            ColumnWithTypeAndName column;
+            column.name = structure_vals[i];
+            column.type = data_type_factory.get(structure_vals[i + 1]);
+            column.column = column.type->createColumn();
+            header.insert(std::move(column));
+        }
+    }
```
Comment (Avogar) [inline programs/obfuscator/Obfuscator.cpp:1339]: I see that we always had such way to parse table structure, but it's incorrect. Type definition can contain commas as well. For example `t Tuple(x UInt32, y String, ...)`.
We have special function to parse structure from string: `parseColumnsListFromString` in `src/TableFunctions/parseColumnsListFromString.h`

File: programs/obfuscator/Obfuscator.cpp [hunk 3]
```diff
@@ -1287,32 +1291,57 @@ try
...
+    if (structure.empty())
     {
-        ColumnWithTypeAndName column;
-        column.name = structure_vals[i];
-        column.type = data_type_factory.get(structure_vals[i + 1]);
-        column.column = column.type->createColumn();
-        header.insert(std::move(column));
+        ReadBufferIterator read_buffer_iterator = [&]()
+        {
+            return std::make_unique<ReadBufferFromFileDescriptor>(STDIN_FILENO);
```
Comment (Avogar) [inline programs/obfuscator/Obfuscator.cpp:1304]: I see you removed using `PeekableReadBuffer`, why? We should come back to the beginning of the data after the schema is inferred

Comment (Avogar) [reply programs/obfuscator/Obfuscator.cpp:1304]: Oh, I see, we expect that stdin is seekable

Comment (evillique) [reply programs/obfuscator/Obfuscator.cpp:1304]: Yep, we check it a bit later. But we can also add the same check with `lseek` here.

**Review by Avogar — APPROVED** [review]

**Pull request merged**